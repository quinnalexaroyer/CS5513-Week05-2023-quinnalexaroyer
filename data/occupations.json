[
{"id":1, "job":"land surveyor"},
{"id":1, "job":"military officer"},
{"id":1, "job":"planter"},
{"id":2, "job":"lawyer"},
{"id":2, "job":"farmer"},
{"id":3, "job":"planter"},
{"id":3, "job":"lawyer"},
{"id":3, "job":"land surveyor"},
{"id":3, "job":"architect"},
{"id":4, "job":"planter"},
{"id":5, "job":"planter"},
{"id":5, "job":"lawyer"},
{"id":6, "job":"lawyer"},
{"id":7, "job":"lawyer"},
{"id":7, "job":"military officer"},
{"id":8, "job":"lawyer"},
{"id":9, "job":"military officer"},
{"id":10, "job":"lawyer"},
{"id":11, "job":"lawyer"},
{"id":11, "job":"planter"},
{"id":12, "job":"military officer"},
{"id":13, "job":"lawyer"},
{"id":14, "job":"lawyer"},
{"id":15, "job":"lawyer"},
{"id":16, "job":"lawyer"},
{"id":16, "job":"land surveyor"},
{"id":17, "job":"tailor"},
{"id":18, "job":"military officer"},
{"id":19, "job":"lawyer"},
{"id":20, "job":"minister"},
{"id":20, "job":"lawyer"},
{"id":20, "job":"teacher"},
{"id":21, "job":"lawyer"},
{"id":21, "job":"teacher"},
{"id":21, "job":"tariff collector"},
{"id":22, "job":"lawyer"},
{"id":23, "job":"court reporter"},
{"id":23, "job":"military officer"},
{"id":24, "job":"lawyer"},
{"id":25, "job":"lawyer"},
{"id":25, "job":"military officer"},
{"id":26, "job":"historian"},
{"id":26, "job":"public servant"},
{"id":26, "job":"naturalist"},
{"id":26, "job":"military officer"},
{"id":26, "job":"police officer"},
{"id":26, "job":"rancher"},
{"id":27, "job":"lawyer"},
{"id":27, "job":"dean"},
{"id":28, "job":"academic"},
{"id":29, "job":"journalist"},
{"id":29, "job":"publisher"},
{"id":30, "job":"lawyer"},
{"id":31, "job":"business person"},
{"id":31, "job":"mining engineer"},
{"id":32, "job":"lawyer"},
{"id":33, "job":"farmer"},
{"id":34, "job":"military officer"},
{"id":34, "job":"university president"},
{"id":35, "job":"journalist"},
{"id":35, "job":"military officer"},
{"id":36, "job":"teacher"},
{"id":36, "job":"military officer"},
{"id":36, "job":"rancher"},
{"id":37, "job":"lawyer"},
{"id":37, "job":"naval officer"},
{"id":38, "job":"lawyer"},
{"id":38, "job":"naval officer"},
{"id":39, "job":"naval officer"},
{"id":39, "job":"farmer"},
{"id":40, "job":"actor"},
{"id":40, "job":"guild president"},
{"id":41, "job":"business person"},
{"id":41, "job":"naval aviator"},
{"id":42, "job":"lawyer"},
{"id":42, "job":"law professor"},
{"id":43, "job":"business person"},
{"id":43, "job":"military pilot"},
{"id":44, "job":"lawyer"},
{"id":44, "job":"law professor"},
{"id":45, "job":"business person"},
{"id":45, "job":"real estate developer"},
{"id":45, "job":"television personality"},
{"id":46, "job":"lawyer"}
]