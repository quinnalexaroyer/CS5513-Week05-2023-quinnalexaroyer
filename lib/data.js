import fs from 'fs';
import path from 'path';
const dataDir = path.join( process.cwd(), 'data' );
export function getSortedList() {
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  let presidentsData = JSON.parse(jsonString);
  presidentsData.sort(
    function(x,y) {
      return x.name.localeCompare(y.name);
    }
  )
  return presidentsData.map(
    function(it) {
      return {id:it.id.toString(), name:it.name};
    }
  )
}
export function getAllIds() {
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const presidentsData = JSON.parse(jsonString);
  return presidentsData.map(
    function(it) {
      return {params: {id:it.id.toString()}};
    }
  )
}
export function getAllStates() {
  const filePath = path.join(dataDir, 'states.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const statesData = JSON.parse(jsonString);
  return Object.keys(statesData).map(
    function(it) {
      return {params: {id:it}};
    }
  );
}
export function getAllParties() {
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const presidentsData = JSON.parse(jsonString);
  let partyList = Array();
  presidentsData.forEach((it) => {
    if(!partyList.includes(it.party)) {
      partyList.push(it.party);
    }
  });
  return partyList.map(
    function(it) {
      return {params: {id:it}};
    }
  );
}
export function getAllOccupations() {
  const filePath = path.join(dataDir, 'occupations.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const occupationData = JSON.parse(jsonString);
  let jobList = Array();
  occupationData.forEach((it) => {
    let slug = makeSlug(it.job);
    if(!jobList.includes(slug)) {
      jobList.push(slug);
    }
  });
  return jobList.map(
    function(it) {
      return {params: {id:it}};
    }
  );
}
function getStateName(abbr) {
  const filePath = path.join(dataDir, 'states.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const stateData = JSON.parse(jsonString);
  if(stateData.hasOwnProperty(abbr)) {
    return stateData[abbr];
  } else {
    return abbr;
  }
}
function getOccupationsForPresident(theID) {
  const filePath = path.join(dataDir, 'occupations.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const occupationData = JSON.parse(jsonString);
  const objectMatch = occupationData.filter(
    function(it) {
      return it.id.toString() == theID.toString();
    }
  );
  return objectMatch;
}
function makeSlug(s) {
  return s.replace(" ", "-");
}
function reverseSlug(s) {
  return s.replace("-", " ");
}
function getOccupationSlugs(occupationObjects) {
  let occupationList = Array();
  occupationObjects.forEach((it) => {
    occupationList.push({job:it.job, slug:makeSlug(it.job)});
  });
  return occupationList;
}
export async function getData(theID) {
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const presidentsData = JSON.parse(jsonString);
  const objectMatch = presidentsData.filter(
    function(it) {
      return it.id.toString() == theID.toString();
    }
  );
  const prevMatch = presidentsData.filter(
    function(thePrev) {
      let prevID = parseInt(theID)-1;
      return thePrev.id.toString() == prevID.toString();
    }
  );
  const nextMatch = presidentsData.filter(
    function(theNext) {
      let nextID = parseInt(theID)+1;
      return theNext.id.toString() == nextID.toString();
    }
  );
  let toReturn;
  if(objectMatch.length > 0) {
    const president = objectMatch[0];
    const stateName = getStateName(president.homestate);
    const occupations = getOccupationSlugs(getOccupationsForPresident(theID));
    let returnPrevMatch;
    if(prevMatch.length > 0) {
      returnPrevMatch = prevMatch[0];
    } else {
      returnPrevMatch = {};
    }
    let returnNextMatch;
    if(nextMatch.length > 0) {
      returnNextMatch = nextMatch[0];
    } else {
      returnNextMatch = {}
    }
    toReturn = {president:objectMatch[0],
                stateName:stateName,
                prev:returnPrevMatch,
                next:returnNextMatch,
                occupations:occupations,
                birthdate:writeDate(objectMatch[0].birthdate),
                deathdate:writeDate(objectMatch[0].deathdate),
                startTerm:writeDate(objectMatch[0].startTerm),
                endTerm:writeDate(objectMatch[0].endTerm)
               };
  } else {
    toReturn = {};
  }
  return toReturn;
}
export async function getDataByState(abbr) {
  const stateName = getStateName(abbr);
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const presidentsData = JSON.parse(jsonString);
  const objectMatch = presidentsData.filter(
    function(it) {
      return it.homestate == abbr;
    }
  );
  return {stateName:stateName, data:objectMatch};
}
export async function getDataByParty(partyName) {
  const filePath = path.join(dataDir, 'presidents.json');
  const jsonString = fs.readFileSync(filePath, 'utf8');
  const presidentsData = JSON.parse(jsonString);
  const objectMatch = presidentsData.filter(
    function(it) {
      return it.party == partyName;
    }
  );
  return objectMatch;
}
export async function getDataByOccupation(slug) {
  const jobFilePath = path.join(dataDir, 'occupations.json');
  const jsonJobString = fs.readFileSync(jobFilePath, 'utf8');
  const occupationData = JSON.parse(jsonJobString);
  const presidentFilePath = path.join(dataDir, 'presidents.json');
  const jsonPresidentString = fs.readFileSync(presidentFilePath, 'utf8');
  const presidentsData = JSON.parse(jsonPresidentString);
  const objectMatch = occupationData.filter(
    function(it) {
      return it.job == reverseSlug(slug);
    }
  );
  let presidentList = Array()
  objectMatch.forEach((it) => {
    presidentList.push(presidentsData.filter(
      function(that) {
        return that.id == it.id;
      }
    )[0])
  });
  return presidentList;
}
function monthName(m) {
  switch(m) {
  case 1:
    return "January";
    break;
  case 2:
    return "February";
    break;
  case 3:
    return "March";
    break;
  case 4:
    return "April";
    break;
  case 5:
    return "May";
    break;
  case 6:
    return "June";
    break;
  case 7:
    return "July";
    break;
  case 8:
    return "August";
    break;
  case 9:
    return "September";
    break;
  case 10:
    return "October";
    break;
  case 11:
    return "November";
    break;
  case 12:
    return "December";
    break;
  }
}
function isDigit(d) {
  return d == '0' || d == '1' || d == '2' || d == '3' || d == '4'
      || d == '5' || d == '6' || d == '7' || d == '8' || d == '9';
}
function writeDate(d) {
  if(  isDigit(d[0])
    && isDigit(d[1])
    && isDigit(d[2])
    && isDigit(d[3])
    && isDigit(d[5])
    && isDigit(d[6])
    && isDigit(d[8])
    && isDigit(d[9])
  ) {
    return (monthName(parseInt(d.slice(5,7)))
          + " " + parseInt(d.slice(8,10)).toString()
          + ", " + parseInt(d.slice(0,4)).toString());
  } else {
    return d;
  }
}