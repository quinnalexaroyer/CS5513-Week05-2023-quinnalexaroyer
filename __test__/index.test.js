import { render, screen } from '@testing-library/react'
import Home from '../pages/index'
import '@testing-library/jest-dom'
import {getSortedList} from '../lib/data';

const presidentsData = getSortedList();

it('renders homepage unchanged', () => {
  const { container } = render(<Home />)
  expect(container).toMatchSnapshot()
})

describe('Home', () => {
  it('renders a list', () => {
    render(<Home presidentsData={presidentsData} />);
    expect(document.querySelector("#listOfPresidents")).toBeTruthy();
    expect(document.querySelector("#listOfPresidents").children.length).toEqual(46);
  })
  it('should fail: test has wrong number of items', () => {
    render(<Home />)
    expect(document.querySelector("#listOfPresidents")).toBeTruthy();
    expect(document.querySelector("#listOfPresidents").children.length).toEqual(32);
  })
})