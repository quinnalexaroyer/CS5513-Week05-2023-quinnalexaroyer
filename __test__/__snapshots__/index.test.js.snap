// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div>
    <header>
      <h1>
        Presidents Data
      </h1>
      <hr />
    </header>
    <main>
      <div
        class="list-group"
        id="listOfPresidents"
      />
    </main>
  </div>
</div>
`;
